pallets/gear/src/tests.rs

```diff
@@ -14830,6 +14830,112 @@ fn critical_hook_in_handle_signal() {
+        // Upload program
+        assert_ok!(Gear::upload_program(
+            RuntimeOrigin::signed(USER_1),
+            WASM_BINARY.to_vec(),
+            DEFAULT_SALT.to_vec(),
+            vec![],
+            10_000_000_000,
+            0,
+            false,
+        ));
+        let pid = get_last_program_id();
+        dbg!("program id", pid);
+
+        run_to_block(2, None);
+
+        assert!(Gear::is_initialized(pid));
+        assert!(utils::is_active(pid));
+
+        // Init conversation
+        assert_ok!(Gear::send_message(
+            RuntimeOrigin::signed(USER_1),
+            pid,
+            EMPTY_PAYLOAD.encode(),
+            10_000_000_000,
+            0,
+            false,
+        ));
+
+        run_to_block(3, None);
+
+        let messages = MailboxOf::<Test>::iter_key(USER_1).map(|(msg, _bn)| msg);
+
+        let mut timeout_msg_id = None;
+
+        for msg in messages {
+            match msg.payload_bytes() {
+                b"for_reply_1" => {
+                    // Reply to the first message
+                    assert_ok!(Gear::send_reply(
+                        RuntimeOrigin::signed(USER_1),
+                        msg.id(),
+                        [1].to_vec(),
+                        1_000_000_000,
+                        0,
+                        false,
+                    ));
+                }
+                b"for_reply_2" => {
+                    // Don't reply, message should time out
+                }
+                b"for_reply_3" => {
+                    // Reply to the third message
+                    assert_ok!(Gear::send_reply(
+                        RuntimeOrigin::signed(USER_1),
+                        msg.id(),
+                        [3].to_vec(),
+                        1_000_000_000,
+                        0,
+                        false,
+                    ));
+                }
+                b"for_reply_4" => {
+                    // reply later
+                    timeout_msg_id = Some(msg.id());
+                }
+                _ => unreachable!(),
+            }
+        }
+
+        run_to_block(4, None);
+        // Expect a reply back
+        assert!(
+            MailboxOf::<Test>::iter_key(USER_1).any(|(m, _)| m.payload_bytes() == b"saw_reply_3")
+        );
+
+        run_to_block(10, None);
+
+        // Program finished
+        assert!(
+            MailboxOf::<Test>::iter_key(USER_1).any(|(m, _)| m.payload_bytes() == b"completed")
+        );
+
+        // Reply to a message that timed out
+        assert_ok!(Gear::send_reply(
+            RuntimeOrigin::signed(USER_1),
+            timeout_msg_id.unwrap(),
+            [4].to_vec(),
+            1_000_000_000,
+            0,
+            false,
+        ));
+        run_to_block(11, None);
+
+        // Hook should still be executed
+        assert!(
+            MailboxOf::<Test>::iter_key(USER_1).any(|(m, _)| m.payload_bytes() == b"saw_reply_4")
+        );
```

Comment:
What's the guarantee this message arrived after the one at line 14918

Reply:
added order check